### DISCUSSION TITLE
Why aren't bv, dd, and aa values stored in population

### DISCUSSION THREAD

**Discussion by gregorgorjanc** [gaynorr/AlphaSimR#75] [General]
AlphaSimR population store genetic value (gv), genotype-by-environment interaction (gxe), and phenotype value, but not the breeding value (bv), dominance deviation (dd), and epistasis deviation (aa).

Why are not bv, dd, and aa stored in the AlphaSimR population?

I am posting here since this will be of interest to many. I am thinking of the design for SIMplyBee where we are mapping individual (multiple!) values to a colony (single!) value, given a user defined function. Its easy with pheno and gv, but I wonder what would be the best design for bv, dd, and aa.

Thanks!

**Comment by gaynorr** [comment]
That's a good question for here. I'll try to answer for a broad audience while hopefully avoiding getting bogged down into too many details. If anyone finds my response confusing, please ask for clarification.

The primary reason is for computational efficiency. Calculating these values take a bit of time and they often aren't needed, so I've set it so that they are only calculated on-demand by calling the `genParam` function directly or indirectly through one of several other functions that rely on it (_e.g._ `bv` or `varA`).

There's also the fact that these values would need to be recalculated anytime the user applied the subset (`[]`) or concatenate (`c()`) functions to a population. This is because an individual's breeding value (bv), dominance deviation (dd) and additive-by-additive epistatic deviation (aa) are all relative values. They are relative to a reference population. In AlphaSimR, the reference population is taken as the population object where an individual resides. This needs to be the reference population if the user wants a meaningful value for a population's additive, dominance or additive-by-additive epistatic genetic variance. This means that if you change the individuals in a population or form a new population every individual needs to have their bv, dd, and aa recalculated. This recalculation doesn't need to happen with phenotypes (pheno) or genetic values (gv), because they aren't relative to a population.

**Reply by gregorgorjanc** [reply]
Brilliant reply @gaynorr!
